File: sdk/ts/src/stream.test.ts
import { test, expect } from "bun:test";
import { SSEStream } from "./stream";

// Live delivery / scope isolation is proven by a real run in the TS live-dashboard example (there is no
// embedded-NATS harness in the TS SDK). Here we cover the pure authorization guards, which need no NATS.

// fakeRes is a minimal ServerResponse stand-in capturing the status and body.
function fakeRes() {
  return {
    statusCode: 0,
    body: "",
    writableLength: 0,
    writeHead(code: number) {
      this.statusCode = code;
    },
    end(s?: string) {
      this.body = s ?? "";
    },
    write() {
      return true;
    },
  };
}

const stubStream = () => new SSEStream({ nats: {} } as any);

test("serveClient rejects an empty scope with 403", async () => {
  const res = fakeRes();
  await stubStream().serveClient({} as any, res as any);
  expect(res.statusCode).toBe(403);
});

test("serveClient rejects a wildcard scope with 400", async () => {
  for (const subj of ["test.>", "test.*.evt"]) {
    const res = fakeRes();
    await stubStream().serveClient({} as any, res as any, subj);
    expect(res.statusCode).toBe(400);
  }
});

File: sdk/python/stream_test.py
"""SSEStream tests for the Python SDK - cover the pure authorization guards, which need no NATS,
mirroring the Go and TS tests. Live delivery / scope isolation is proven by a real run in the Python
live-dashboard example.

Run: uv run --with nats-py --with aiohttp -m unittest stream_test   (from sdk/python)
"""

import unittest

from aether import Ctx, SSEStream


def _stub_stream() -> SSEStream:
    # nats is never touched: the guards under test fire before any subscribe.
    return SSEStream(Ctx(nats=None, name="edge", app="test"))


class SSEStreamTest(unittest.IsolatedAsyncioTestCase):
    async def test_empty_scope_is_403(self):
        resp = await _stub_stream().serve_client(None)
        self.assertEqual(resp.status, 403)

    async def test_wildcard_scope_is_400(self):
        for subj in ("test.>", "test.*.evt"):
            resp = await _stub_stream().serve_client(None, subj)
            self.assertEqual(resp.status, 400)


if __name__ == "__main__":
    unittest.main()
